File: py/wlearn/preprocess.py
"""ML preprocessing: imputation, encoding, scaling.

Learns parameters during fit(), applies during transform().
State is serializable for .wlrn bundles.
"""

import math
import json

import numpy as np

from .errors import ValidationError


class Preprocessor:
    """Automatic feature preprocessing for tabular data.

    Handles missing values, categorical encoding, and feature scaling.
    Learns parameters from training data and applies them consistently
    at inference time.
    """

    def __init__(self, impute='auto', encode='auto', scale=False,
                 max_categories=20):
        self._config = {
            'impute': impute,
            'encode': encode,
            'scale': scale,
            'maxCategories': max_categories,
        }
        self._fitted = False
        self._col_types = None
        self._impute_values = None
        self._encodings = None
        self._scale_params = None
        self._output_cols = 0

    def fit(self, X, y=None):
        """Learn preprocessing parameters from training data."""
        X = np.asarray(X, dtype=np.float64)
        rows, cols = X.shape

        self._col_types = [None] * cols
        self._impute_values = np.zeros(cols, dtype=np.float64)
        col_stats = [None] * cols

        for c in range(cols):
            col = X[:, c]
            mask = ~np.isnan(col)
            values = col[mask]
            has_nan = not mask.all()

            unique = set(values.tolist())
            is_integer = all(v == int(v) for v in values)
            is_categorical = (
                self._config['encode'] is not False and
                is_integer and
                len(unique) <= self._config['maxCategories'] and
                len(unique) >= 2
            )

            self._col_types[c] = 'categorical' if is_categorical else 'numeric'

            if self._config['impute'] is not False and has_nan:
                if is_categorical:
                    # Mode
                    from collections import Counter
                    counter = Counter(values.tolist())
                    self._impute_values[c] = counter.most_common(1)[0][0]
                else:
                    # Mean
                    self._impute_values[c] = values.mean() if len(values) > 0 else 0.0

            col_stats[c] = {
                'values': values,
                'unique': unique,
                'sum': float(values.sum()) if len(values) > 0 else 0.0,
                'count': len(values),
            }

        # Build encodings
        self._encodings = [None] * cols
        output_idx = 0

        if self._config['encode'] is not False:
            for c in range(cols):
                if self._col_types[c] == 'categorical':
                    sorted_vals = sorted(col_stats[c]['unique'])
                    if self._config['encode'] == 'label':
                        mapping = {v: i for i, v in enumerate(sorted_vals)}
                        self._encodings[c] = {
                            'type': 'label',
                            'mapping': mapping,
                            'startIdx': output_idx,
                        }
                        output_idx += 1
                    else:
                        mapping = {v: output_idx + i
                                   for i, v in enumerate(sorted_vals)}
                        self._encodings[c] = {
                            'type': 'onehot',
                            'mapping': mapping,
                            'startIdx': output_idx,
                            'size': len(sorted_vals),
                        }
                        output_idx += len(sorted_vals)
                else:
                    output_idx += 1
        else:
            output_idx = cols

        self._output_cols = output_idx

        # Compute scaling parameters
        self._scale_params = [None] * self._output_cols

        if self._config['scale']:
            out_c = 0
            for c in range(cols):
                if (self._col_types[c] == 'categorical' and
                        self._config['encode'] is not False):
                    enc = self._encodings[c]
                    out_c += enc['size'] if enc['type'] == 'onehot' else 1
                    continue
                vals = col_stats[c]['values']
                if self._config['scale'] == 'standard':
                    mean = float(vals.mean()) if len(vals) > 0 else 0.0
                    std = float(vals.std()) if len(vals) > 0 else 1.0
                    self._scale_params[out_c] = {
                        'mean': mean,
                        'std': std if std > 0 else 1.0,
                    }
                elif self._config['scale'] == 'minmax':
                    mn = float(vals.min()) if len(vals) > 0 else 0.0
                    mx = float(vals.max()) if len(vals) > 0 else 1.0
                    rng = mx - mn
                    self._scale_params[out_c] = {
                        'min': mn,
                        'range': rng if rng > 0 else 1.0,
                    }
                out_c += 1

        self._fitted = True
        return self

    def transform(self, X):
        """Apply learned preprocessing to new data."""
        if not self._fitted:
            raise ValidationError('Preprocessor not fitted')
        X = np.asarray(X, dtype=np.float64)
        rows, cols = X.shape
        out_cols = self._output_cols
        out = np.zeros((rows, out_cols), dtype=np.float64)

        for r in range(rows):
            out_c = 0
            for c in range(cols):
                v = X[r, c]

                # Impute missing
                if math.isnan(v) and self._config['impute'] is not False:
                    v = self._impute_values[c]

                # Encode
                if (self._col_types[c] == 'categorical' and
                        self._encodings[c] is not None):
                    enc = self._encodings[c]
                    if enc['type'] == 'onehot':
                        idx = enc['mapping'].get(v)
                        if idx is not None:
                            out[r, idx] = 1.0
                        out_c = enc['startIdx'] + enc['size']
                    else:
                        label = enc['mapping'].get(v, -1)
                        out[r, out_c] = label
                        out_c += 1
                else:
                    out[r, out_c] = v
                    out_c += 1

        # Apply scaling
        if self._config['scale']:
            for j in range(out_cols):
                sp = self._scale_params[j]
                if sp is None:
                    continue
                if self._config['scale'] == 'standard':
                    out[:, j] = (out[:, j] - sp['mean']) / sp['std']
                elif self._config['scale'] == 'minmax':
                    out[:, j] = (out[:, j] - sp['min']) / sp['range']

        return out

    def fit_transform(self, X, y=None):
        """Fit and transform in one call."""
        self.fit(X, y)
        return self.transform(X)

    def get_state(self):
        """Serialize fitted state."""
        if not self._fitted:
            raise ValidationError('Preprocessor not fitted')
        return {
            'config': self._config,
            'colTypes': self._col_types,
            'imputeValues': self._impute_values.tolist(),
            'encodings': [
                {
                    'type': e['type'],
                    'mapping': list(e['mapping'].items()),
                    'startIdx': e['startIdx'],
                    'size': e.get('size'),
                } if e else None
                for e in self._encodings
            ],
            'scaleParams': self._scale_params,
            'outputCols': self._output_cols,
        }

    @classmethod
    def from_state(cls, state):
        """Restore from serialized state."""
        pp = cls(**{
            'impute': state['config']['impute'],
            'encode': state['config']['encode'],
            'scale': state['config']['scale'],
            'max_categories': state['config']['maxCategories'],
        })
        pp._col_types = state['colTypes']
        pp._impute_values = np.array(state['imputeValues'], dtype=np.float64)
        pp._encodings = [
            {
                'type': e['type'],
                'mapping': {k: v for k, v in e['mapping']},
                'startIdx': e['startIdx'],
                'size': e.get('size'),
            } if e else None
            for e in state['encodings']
        ]
        pp._scale_params = state['scaleParams']
        pp._output_cols = state['outputCols']
        pp._fitted = True
        return pp

    @property
    def is_fitted(self):
        return self._fitted

    @property
    def output_cols(self):
        return self._output_cols

    def dispose(self):
        pass

File: packages/core/src/preprocess.js
const { normalizeX, normalizeY } = require('./matrix.js')
const { ValidationError } = require('./errors.js')

/**
 * ML preprocessing: imputation, encoding, scaling.
 * Learns parameters during fit(), applies during transform().
 * State is serializable for .wlrn bundles.
 *
 * Designed to be used as a Pipeline step before model fitting.
 * Future: delegate to tranfi transforms once fitted state serialization lands.
 */
class Preprocessor {
  #config
  #fitted = false
  // Learned state
  #colTypes = null     // 'numeric' | 'categorical' per column
  #imputeValues = null // per-column impute value (mean for numeric, mode for categorical)
  #encodings = null    // per-column: { type: 'onehot', mapping: { val: colIdx } }
  #scaleParams = null  // per-column: { mean, std } or { min, range }
  #outputCols = 0      // total output columns after encoding

  constructor(config = {}) {
    this.#config = {
      impute: config.impute ?? 'auto',          // 'auto' | 'mean' | 'median' | 'zero' | false
      encode: config.encode ?? 'auto',          // 'auto' | 'onehot' | 'label' | false
      scale: config.scale ?? false,             // 'standard' | 'minmax' | false
      maxCategories: config.maxCategories ?? 20, // columns with more unique values treated as numeric
      ...config,
    }
  }

  /**
   * Learn preprocessing parameters from training data.
   * X is a DenseMatrix { data, rows, cols }.
   */
  fit(X, y) {
    const Xn = normalizeX(X)
    const { data, rows, cols } = Xn

    // Step 1: Detect column types and compute statistics
    this.#colTypes = new Array(cols)
    this.#imputeValues = new Float64Array(cols)
    const colStats = new Array(cols)

    for (let c = 0; c < cols; c++) {
      const values = []
      let hasNaN = false
      let sum = 0
      let count = 0

      for (let r = 0; r < rows; r++) {
        const v = data[r * cols + c]
        if (Number.isNaN(v)) {
          hasNaN = true
        } else {
          values.push(v)
          sum += v
          count++
        }
      }

      // Detect categorical: integer-valued with few unique values
      const unique = new Set(values)
      const isInteger = values.every(v => v === Math.floor(v))
      const isCategorical = this.#config.encode !== false &&
        isInteger && unique.size <= this.#config.maxCategories && unique.size >= 2

      this.#colTypes[c] = isCategorical ? 'categorical' : 'numeric'

      // Compute impute value
      if (this.#config.impute !== false && hasNaN) {
        if (isCategorical) {
          // Mode
          const freq = new Map()
          for (const v of values) freq.set(v, (freq.get(v) || 0) + 1)
          let modeVal = values[0], modeCount = 0
          for (const [v, ct] of freq) {
            if (ct > modeCount) { modeVal = v; modeCount = ct }
          }
          this.#imputeValues[c] = modeVal
        } else {
          // Mean
          this.#imputeValues[c] = count > 0 ? sum / count : 0
        }
      }

      colStats[c] = { values, unique, sum, count }
    }

    // Step 2: Build encodings for categorical columns
    this.#encodings = new Array(cols).fill(null)
    let outputIdx = 0

    if (this.#config.encode !== false) {
      for (let c = 0; c < cols; c++) {
        if (this.#colTypes[c] === 'categorical') {
          const sorted = [...colStats[c].unique].sort((a, b) => a - b)
          if (this.#config.encode === 'label' || this.#config.encode === false) {
            // Label encoding: 1 output column
            const mapping = new Map()
            sorted.forEach((v, i) => mapping.set(v, i))
            this.#encodings[c] = { type: 'label', mapping, startIdx: outputIdx }
            outputIdx++
          } else {
            // One-hot encoding (default for 'auto' and 'onehot')
            const mapping = new Map()
            sorted.forEach((v, i) => mapping.set(v, outputIdx + i))
            this.#encodings[c] = { type: 'onehot', mapping, startIdx: outputIdx, size: sorted.length }
            outputIdx += sorted.length
          }
        } else {
          outputIdx++
        }
      }
    } else {
      outputIdx = cols
    }
    this.#outputCols = outputIdx

    // Step 3: Compute scaling parameters (on non-NaN values)
    this.#scaleParams = new Array(this.#outputCols).fill(null)

    if (this.#config.scale) {
      // Recompute on the output space after encoding
      // For now, scale numeric columns only
      let outC = 0
      for (let c = 0; c < cols; c++) {
        if (this.#colTypes[c] === 'categorical' && this.#config.encode !== false) {
          const enc = this.#encodings[c]
          outC += enc.type === 'onehot' ? enc.size : 1
          continue
        }
        const vals = colStats[c].values
        if (this.#config.scale === 'standard') {
          const mean = colStats[c].sum / (colStats[c].count || 1)
          let variance = 0
          for (const v of vals) variance += (v - mean) ** 2
          variance /= (vals.length || 1)
          const std = Math.sqrt(variance)
          this.#scaleParams[outC] = { mean, std: std || 1 }
        } else if (this.#config.scale === 'minmax') {
          let min = Infinity, max = -Infinity
          for (const v of vals) { if (v < min) min = v; if (v > max) max = v }
          const range = max - min || 1
          this.#scaleParams[outC] = { min, range }
        }
        outC++
      }
    }

    this.#fitted = true
    return this
  }

  /**
   * Apply learned preprocessing to new data.
   * Returns a new DenseMatrix.
   */
  transform(X) {
    if (!this.#fitted) throw new ValidationError('Preprocessor not fitted')
    const Xn = normalizeX(X)
    const { data, rows, cols } = Xn
    const outCols = this.#outputCols
    const out = new Float64Array(rows * outCols)

    for (let r = 0; r < rows; r++) {
      let outC = 0
      for (let c = 0; c < cols; c++) {
        let v = data[r * cols + c]

        // Impute missing
        if (Number.isNaN(v) && this.#config.impute !== false) {
          v = this.#imputeValues[c]
        }

        // Encode
        if (this.#colTypes[c] === 'categorical' && this.#encodings[c]) {
          const enc = this.#encodings[c]
          if (enc.type === 'onehot') {
            // Zero-fill the one-hot range
            const idx = enc.mapping.get(v)
            if (idx !== undefined) {
              out[r * outCols + idx] = 1
            }
            outC = enc.startIdx + enc.size
          } else {
            // Label encoding
            const label = enc.mapping.get(v)
            out[r * outCols + outC] = label !== undefined ? label : -1
            outC++
          }
        } else {
          out[r * outCols + outC] = v
          outC++
        }
      }

      // Apply scaling
      if (this.#config.scale) {
        for (let j = 0; j < outCols; j++) {
          const sp = this.#scaleParams[j]
          if (!sp) continue
          if (this.#config.scale === 'standard') {
            out[r * outCols + j] = (out[r * outCols + j] - sp.mean) / sp.std
          } else if (this.#config.scale === 'minmax') {
            out[r * outCols + j] = (out[r * outCols + j] - sp.min) / sp.range
          }
        }
      }
    }

    return { data: out, rows, cols: outCols }
  }

  /**
   * Fit and transform in one call.
   */
  fitTransform(X, y) {
    this.fit(X, y)
    return this.transform(X)
  }

  /**
   * Serialize fitted state for .wlrn bundle.
   */
  getState() {
    if (!this.#fitted) throw new ValidationError('Preprocessor not fitted')
    return {
      config: this.#config,
      colTypes: this.#colTypes,
      imputeValues: [...this.#imputeValues],
      encodings: this.#encodings.map(e => {
        if (!e) return null
        return {
          type: e.type,
          mapping: [...e.mapping.entries()],
          startIdx: e.startIdx,
          size: e.size,
        }
      }),
      scaleParams: this.#scaleParams,
      outputCols: this.#outputCols,
    }
  }

  /**
   * Restore from serialized state.
   */
  static fromState(state) {
    const pp = new Preprocessor(state.config)
    pp.#colTypes = state.colTypes
    pp.#imputeValues = new Float64Array(state.imputeValues)
    pp.#encodings = state.encodings.map(e => {
      if (!e) return null
      return {
        type: e.type,
        mapping: new Map(e.mapping),
        startIdx: e.startIdx,
        size: e.size,
      }
    })
    pp.#scaleParams = state.scaleParams
    pp.#outputCols = state.outputCols
    pp.#fitted = true
    return pp
  }

  get isFitted() { return this.#fitted }
  get outputCols() { return this.#outputCols }

  get capabilities() {
    return { transformer: true }
  }

  dispose() {}
}

module.exports = { Preprocessor }
